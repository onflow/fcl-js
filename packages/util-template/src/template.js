export function interleave(a = [], b = [], c = []) {
  if (!a.length && !b.length) return c
  if (!a.length) return c
  if (!b.length) return [...c, a[0]]

  const [aHead, ...aRest] = a
  const [bHead, ...bRest] = b

  if (aHead !== undefined) c.push(aHead)
  if (bHead !== undefined) c.push(bHead)

  return interleave(aRest, bRest, c)
}

function recApply(d) {
  return function(arg1) {
    if (typeof arg1 === "function") {
      console.warn(
        `
        %cFCL/SDK Deprecation Notice
        ============================

        Interopolation of functions into template literals will not be a thing in future versions of the Flow-JS-SDK or FCL.
        You can learn more (including a guide on common transition paths) here: https://github.com/onflow/flow-js-sdk/blob/master/packages/sdk/TRANSITIONS.md#0001-deprecate-params

        ============================
      `,
        "font-weight:bold;font-family:monospace;"
      )
      return recApply(d)(arg1(d))
    }
    return String(arg1)
  }
}

export function template(head, ...rest) {
  if (typeof head === "string") return () => head
  if (Array.isArray(head)) {
    return d =>
      interleave(head, rest.map(recApply(d)))
        .join("")
        .trim()
  }
  return head
}
