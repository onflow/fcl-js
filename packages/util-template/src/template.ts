import {log} from "@onflow/util-logger"

/**
 * Interleaves two arrays
 * @param a - The first array
 * @param b - The second array
 * @param c - The target array
 * @returns The interleaved array
 */
export function interleave<A, B>(
  a: A[] = [],
  b: B[] = [],
  c: (A | B)[] = []
): (A | B)[] {
  if (!a.length && !b.length) return c
  if (!a.length) return c
  if (!b.length) {
    c.push(...a)
    return c
  }

  const [aHead, ...aRest] = a
  const [bHead, ...bRest] = b

  if (aHead !== undefined) c.push(aHead)
  if (bHead !== undefined) c.push(bHead)

  return interleave(aRest, bRest, c)
}

/**
 * Recursively apply a value to a function
 * @param d - The value to apply
 * @returns A function that takes a function and applies the value to it
 */
function recApply<T, U>(d: T): (x: U) => string {
  return function (arg1) {
    if (typeof arg1 === "function") {
      log.deprecate({
        pkg: "FCL/SDK",
        subject: "Interopolation of functions into template literals",
        transition:
          "https://github.com/onflow/flow-js-sdk/blob/master/packages/sdk/TRANSITIONS.md#0001-deprecate-params",
      })
      return recApply(d)(arg1(d))
    }
    return String(arg1)
  }
}

/**
 * Creates a template function
 * @param head - A string, template string array, or template function
 * @param rest - The rest of the arguments
 * @returns A template function
 */
export function template(
  head: string | TemplateStringsArray | ((x?: unknown) => string),
  ...rest: unknown[]
): (x?: unknown) => string {
  if (typeof head === "string") return () => head
  if (typeof head === "function") return head
  return (x: unknown) =>
    interleave([...head], rest.map(recApply(x)))
      .join("")
      .trim()
}
