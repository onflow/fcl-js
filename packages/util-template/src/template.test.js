import {template, interleave} from "./template"

describe("interleave", () => {
  ;[
    [[], [], []],
    [[1], [], [1]],
    [[], [1], []],
    [[1], [2], [1, 2]],
    [
      [1, 2, 3],
      [4, 5, 6],
      [1, 4, 2, 5, 3, 6],
    ],
    [
      [1, 2, 3],
      [4, 5],
      [1, 4, 2, 5, 3],
    ],
  ].forEach(([a, b, c]) => {
    test(`interleave([${a}], [${b}]) -> [${c}]`, () => {
      expect(interleave(a, b)).toStrictEqual(c)
    })
  })
})

const _ = (msg, a, b) => {
  if (b == null) {
    test(msg, () => expect(a).toMatchSnapshot())
  } else {
    test(msg, () => expect(a).toBe(b))
  }
}
const t = v => typeof v

describe("template", () => {
  describe("input type vs output type", () => {
    _("template`` -> function", t(template``))
    _("t71(``) -> function", t(template("words")))
    _("template(template``) -> function", t(template(template``)))
  })

  describe("no interop", () => {
    _("template`abc`() -> 'abc'", template`abc`())
    _("template('abc')() -> 'abc'", template("abc")())
    _("template(template`abc`)() -> 'abc'", template(template`abc`)())
  })

  describe("only interop", () => {
    _("template`${'abc'}`() -> 'abc'", template`${"abc"}`())
    _("template(template`${'abc'}`)() -> 'abc'", template(template`${"abc"}`)())
  })

  describe("interop function", () => {
    const o = {a: "abc"}
    _("template`${o=>o.a}`(o) -> 'abc'", template`${o => o.a}`(o))
    _(
      "template(template`${o=>o.a}`)(o) -> 'abc'",
      template(template`${o => o.a}`)(o)
    )
  })

  describe("interop more template", () => {
    const o = {a: "abc"}
    _(
      "template`x${template`y${o=>o.a}`}`(o) => 'xyabc'",
      template`x${template`y${o => o.a}`}`(o)
    )
    _(
      "template`x${template`y${template`z${o=>o.a}`}`}`(o) => 'xyabc'",
      template`x${template`y${template`z${o => o.a}`}`}`(o)
    )
    _(
      "template(template`x${template`y${o => o.a}`}`)(o) -> 'xyabc'",
      template(template`x${template`y${o => o.a}`}`)(o)
    )
  })

  describe("interop nested functions", () => {
    const fn = a => b => c => d => e => f => f.a
    const o = {a: "abc"}
    _("template`${fn}`(o) -> 'abc'", template`${fn}`(o))
  })

  describe("template takes a string", () => {
    const str = "woot woot im a boot"
    _("template(str)() -> str", template(str)(), str)
  })

  describe("object can have non string values", () => {
    const data = {a: 1, b: NaN, c: undefined, d: null, e: false, f: true}
    const tx = template`a:${o => o.a}|b:${o => o.b}|c:${o => o.c}|d:${o =>
      o.d}|e:${o => o.e}|f:${o => o.f}`
    _("template(data)", tx(data), "a:1|b:NaN|c:undefined|d:null|e:false|f:true")
  })
})
