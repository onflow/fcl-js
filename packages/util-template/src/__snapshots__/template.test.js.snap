// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template input type vs output type t71(\`\`) -> function 1`] = `"function"`;

exports[`template input type vs output type template(template\`\`) -> function 1`] = `"function"`;

exports[`template input type vs output type template\`\` -> function 1`] = `"function"`;

exports[`template interop function template(template\`\${o=>o.a}\`)(o) -> 'abc' 1`] = `"abc"`;

exports[`template interop function template\`\${o=>o.a}\`(o) -> 'abc' 1`] = `"abc"`;

exports[`template interop more template template(template\`x\${template\`y\${o => o.a}\`}\`)(o) -> 'xyabc' 1`] = `"xyabc"`;

exports[`template interop more template template\`x\${template\`y\${o=>o.a}\`}\`(o) => 'xyabc' 1`] = `"xyabc"`;

exports[`template interop more template template\`x\${template\`y\${template\`z\${o=>o.a}\`}\`}\`(o) => 'xyabc' 1`] = `"xyzabc"`;

exports[`template interop nested functions template\`\${fn}\`(o) -> 'abc' 1`] = `"abc"`;

exports[`template no interop template('abc')() -> 'abc' 1`] = `"abc"`;

exports[`template no interop template(template\`abc\`)() -> 'abc' 1`] = `"abc"`;

exports[`template no interop template\`abc\`() -> 'abc' 1`] = `"abc"`;

exports[`template only interop template(template\`\${'abc'}\`)() -> 'abc' 1`] = `"abc"`;

exports[`template only interop template\`\${'abc'}\`() -> 'abc' 1`] = `"abc"`;
